File: client-sdk/ios/WebRTCKit/Presentation/Views/QualityMetricsOverlay.swift
// MARK: - Quality Metrics Overlay
// TRUST 5 Compliance: Testable, Unified, Trackable

import SwiftUI
import Combine

// MARK: - Quality Metrics Overlay View

/// Overlay component for displaying WebRTC quality metrics
/// Shows RTT, packet loss, bitrate, and quality state with color coding
@available(iOS 13.0, *)
struct QualityMetricsOverlay: View {

    // MARK: - Properties

    let statsReport: RTCStatsReport?
    let isVisible: Bool

    // MARK: - Body

    var body: some View {
        if isVisible, let report = statsReport {
            VStack(alignment: .leading, spacing: 8) {
                // Quality State Header
                Text("Connection Quality: \(report.getQualityState().displayName)")
                    .font(.headline)
                    .foregroundColor(.white)

                Divider()
                    .background(Color.white.opacity(0.3))

                // RTT Metric
                MetricRow(
                    label: "RTT",
                    value: "\(Int(report.rtt)) ms",
                    color: getMetricColor(
                        value: report.rtt,
                        threshold: 100.0,
                        lowerIsBetter: true
                    )
                )

                // Packet Loss Metric
                MetricRow(
                    label: "Packet Loss",
                    value: String(format: "%.2f%%", report.packetLoss),
                    color: getMetricColor(
                        value: report.packetLoss,
                        threshold: 3.0,
                        lowerIsBetter: true
                    )
                )

                // Bitrate Metric
                MetricRow(
                    label: "Bitrate",
                    value: formatBitrate(bps: report.bitrate),
                    color: getMetricColor(
                        value: report.bitrate,
                        threshold: 500_000.0,
                        lowerIsBetter: false
                    )
                )

                // Resolution Metric
                if report.resolutionWidth > 0 && report.resolutionHeight > 0 {
                    MetricRow(
                        label: "Resolution",
                        value: "\(report.resolutionWidth)x\(report.resolutionHeight)",
                        color: .white
                    )
                }

                Divider()
                    .background(Color.white.opacity(0.3))

                // Quality Score
                Text("Quality Score: \(report.calculateQualityScore())/100")
                    .font(.caption)
                    .foregroundColor(.white.opacity(0.8))
            }
            .padding()
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(Color(hex: report.getQualityState().color).opacity(0.9))
            )
            .padding()
        }
    }

    // MARK: - Private Methods

    private func getMetricColor(value: Double, threshold: Double, lowerIsBetter: Bool) -> Color {
        if lowerIsBetter {
            switch value {
            case 0..<threshold * 0.5:
                return .green
            case threshold * 0.5..<threshold:
                return .orange
            default:
                return .red
            }
        } else {
            switch value {
            case threshold * 2...:
                return .green
            case threshold..<threshold * 2:
                return .orange
            default:
                return .red
            }
        }
    }

    private func formatBitrate(bps: Double) -> String {
        switch bps {
        case 1_000_000...:
            return String(format: "%.2f Mbps", bps / 1_000_000)
        case 1_000...:
            return String(format: "%.2f Kbps", bps / 1_000)
        default:
            return String(format: "%.2f bps", bps)
        }
    }
}

// MARK: - Metric Row View

@available(iOS 13.0, *)
struct MetricRow: View {
    let label: String
    let value: String
    let color: Color

    var body: some View {
        HStack {
            Text(label)
                .foregroundColor(.white)
            Spacer()
            Text(value)
                .foregroundColor(color)
                .bold()
        }
    }
}

// MARK: - Color Extension

extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (1, 1, 1, 0)
        }

        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue:  Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}

// MARK: - Preview

@available(iOS 13.0, *)
struct QualityMetricsOverlay_Previews: PreviewProvider {
    static var previews: some View {
        QualityMetricsOverlay(
            statsReport: RTCStatsReport(
                timestamp: Date().timeIntervalSince1970 * 1000,
                rtt: 45.0,
                packetLoss: 0.5,
                bitrate: 1_500_000.0,
                resolutionWidth: 1280,
                resolutionHeight: 720,
                bytesReceived: 10_000_000,
                bytesSent: 5_000_000,
                framesDecoded: 3000,
                framesEncoded: 3000
            ),
            isVisible: true
        )
    }
}

File: client-sdk/android/webrtc-core/src/main/java/com/webrtclite/core/presentation/ui/QualityMetricsOverlay.kt
package com.webrtclite.core.presentation.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.webrtclite.core.webrtc.QualityState
import com.webrtclite.core.webrtc.RTCStatsReport

/**
 * Overlay component for displaying WebRTC quality metrics
 * Shows RTT, packet loss, bitrate, and quality state with color coding
 */
@Composable
fun QualityMetricsOverlay(
    statsReport: RTCStatsReport?,
    isVisible: Boolean,
    modifier: Modifier = Modifier
) {
    if (!isVisible || statsReport == null) {
        return
    }

    val qualityState = statsReport.getQualityState()
    val backgroundColor = getQualityColor(qualityState).copy(alpha = 0.9f)

    Box(
        modifier = modifier
            .fillMaxWidth()
            .padding(16.dp)
            .background(
                color = backgroundColor,
                shape = RoundedCornerShape(12.dp)
            )
            .padding(16.dp)
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp),
            horizontalAlignment = Alignment.Start
        ) {
            // Quality State Header
            Text(
                text = "Connection Quality: ${qualityState.name}",
                style = MaterialTheme.typography.titleMedium,
                color = Color.White,
                fontSize = 18.sp
            )

            Spacer(modifier = Modifier.height(4.dp))

            // RTT Metric
            MetricRow(
                label = "RTT",
                value = "${statsReport.rtt.toInt()} ms",
                color = getMetricColor(statsReport.rtt, 100.0)
            )

            // Packet Loss Metric
            MetricRow(
                label = "Packet Loss",
                value = "${"%.2f".format(statsReport.packetLoss)}%",
                color = getMetricColor(statsReport.packetLoss, 3.0)
            )

            // Bitrate Metric
            MetricRow(
                label = "Bitrate",
                value = formatBitrate(statsReport.bitrate),
                color = getMetricColorInverse(statsReport.bitrate, 500_000.0)
            )

            // Resolution Metric
            if (statsReport.resolutionWidth > 0 && statsReport.resolutionHeight > 0) {
                MetricRow(
                    label = "Resolution",
                    value = "${statsReport.resolutionWidth}x${statsReport.resolutionHeight}",
                    color = Color.White
                )
            }

            // Quality Score
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                text = "Quality Score: ${statsReport.calculateQualityScore()}/100",
                style = MaterialTheme.typography.bodySmall,
                color = Color.White.copy(alpha = 0.8f)
            )
        }
    }
}

@Composable
private fun MetricRow(
    label: String,
    value: String,
    color: Color
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = Color.White
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodyMedium,
            color = color
        )
    }
}

/**
 * Get color based on quality state
 */
private fun getQualityColor(state: QualityState): Color {
    return when (state) {
        QualityState.EXCELLENT -> Color(0xFF4CAF50)  // Green
        QualityState.GOOD -> Color(0xFF8BC34A)       // Light Green
        QualityState.FAIR -> Color(0xFFFF9800)       // Orange
        QualityState.POOR -> Color(0xFFF44336)       // Red
    }
}

/**
 * Get metric color (lower is better for RTT and packet loss)
 */
private fun getMetricColor(value: Double, threshold: Double): Color {
    return when {
        value < threshold * 0.5 -> Color(0xFF4CAF50)  // Green
        value < threshold -> Color(0xFFFF9800)        // Orange
        else -> Color(0xFFF44336)                     // Red
    }
}

/**
 * Get metric color (higher is better for bitrate)
 */
private fun getMetricColorInverse(value: Double, threshold: Double): Color {
    return when {
        value > threshold * 2 -> Color(0xFF4CAF50)    // Green
        value > threshold -> Color(0xFFFF9800)        // Orange
        else -> Color(0xFFF44336)                     // Red
    }
}

/**
 * Format bitrate for display
 */
private fun formatBitrate(bps: Double): String {
    return when {
        bps >= 1_000_000 -> String.format("%.2f Mbps", bps / 1_000_000)
        bps >= 1_000 -> String.format("%.2f Kbps", bps / 1_000)
        else -> String.format("%.2f bps", bps)
    }
}
